framework/local-cryptographic-materials-cache.md

```diff
@@ -0,0 +1,104 @@
+| Javascript | 0.1.0                                  | 0.1.0                     | [get_local_cryptographic_materials_cache.ts](https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/cache-material/src/get_local_cryptographic_materials_cache.ts) |
+| Python     | n/a                                    | n/a                       | [caches/local.py](https://github.com/aws/aws-encryption-sdk-python/blob/master/src/aws_encryption_sdk/caches/local.py)                                                               |
+| Java       | 0.1.0                                  | 1.3.0                     | [LocalCryptoMaterialsCache.java](https://github.com/aws/aws-encryption-sdk-java/blob/master/src/main/java/com/amazonaws/encryptionsdk/caching/LocalCryptoMaterialsCache.java)        |
+
+## Overview
+
+The local Cryptographic Materials Cache (local CMC)
+is a built-in implementation of the [CMC interface](cryptographic-materials-cache.md)
+provided by the AWS Encryption SDK.
+The local CMC is a configurable, in-memory, least recently used (LRU) cache.
+It provides non-blocking, locking, [cache entries](cryptographic-materials-cache.md#cache-entry)
+per [cache identifier](cryptographic-materials-cache.md#cache-identifier).
+
+## Definitions
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Time-to-live (TTL)
+
+Each cache entry has a time-to-live (TTL) field
+that represents a point in time at which the cache entry MUST be considered invalid.
+After a cache entry's TTL has elapsed,
+we say that the entry is _TTL-expired_,
+and the local CMC MUST NOT return the entry to any user.
+
+## Initialization
+
+On initialization of the local CMC,
+the user MUST provide the following:
+
+- [Entry Capacity](#entry-capacity)
+
+The local CMC MUST also define the following:
+
+- [Entry Pruning Tail Size](#entry-pruning-tail-size)
+
+### Entry Capacity
+
+The _entry capacity_ is the maximum size of the local CMC in terms of cache entries.
+The local CMC MUST NOT store more entries than this value,
+except temporarily while performing a Put Cache Entry operation.
+The local CMC MUST accept entry capacity values between zero
+and an implementation-defined maximum, inclusive.
+
+### Entry Pruning Tail Size
+
+The _entry pruning tail size_
+is the number of least recently used entries that the local CMC
+MUST check during [pruning](#pruning)
+for TTL-expired entries to evict.
+
+## Behaviors
+
+### Put Cache Entry
+
+When calling the Put Cache Entry operation,
+the user MUST provide a time-to-live (TTL) value.
```

Comment:
So, this goes a bit against `a point in time` from above.

Are customers providing a time delta (like 5 seconds) or are they providing a timestamp for when they want the element to expire. I would recommend this value specifically be a delta, so we don't have to worry about different timedate formats, time zones, etc.

Internally, we can store it however we want, but we should have customers provide the difference in time, not "a point in time" (a timestamp) like we define TTL higher up.

Reply:
I intentionally define TTL abstractly as "represent[ing] a point in time", rather than as "a time delta" or "a timestamp", so that implementations have the freedom to choose both their internal representation and the representation of the TTL parameter in Put Cache Entry.

I'm hesitant to _require_ that implementations accept the TTL parameter as a time delta _and not_ as a timestamp, since I think individual implementations are in the best position to decide what's right for them.

Reply:
I pretty strongly think accepting timestamps will lead to implementation errors or potential customer confusion, but I'll leave this up to you/the rest of the team.

Reply:
That's fair. I'll change the wording to explain that the caller provide a time delta specifically, whereas the local CMC can choose how to derive the TTL value.